src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    internal class HostingApplicationDiagnostics
+    {
+        private static readonly double TimestampToTicks = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+        private const string ActivityName = "Microsoft.AspNetCore.Hosting.HttpRequestIn";
+        private const string ActivityStartKey = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start";
+
+        private const string DiagnosticsBeginRequestKey = "Microsoft.AspNetCore.Hosting.BeginRequest";
+        private const string DiagnosticsEndRequestKey = "Microsoft.AspNetCore.Hosting.EndRequest";
+        private const string DiagnosticsUnhandledExceptionKey = "Microsoft.AspNetCore.Hosting.UnhandledException";
+
+        private const string RequestIdHeaderName = "Request-Id";
+        private const string CorrelationContextHeaderName = "Correlation-Context";
+
+        private readonly DiagnosticListener _diagnosticListener;
+        private readonly ILogger _logger;
+
+        public HostingApplicationDiagnostics(ILogger logger, DiagnosticListener diagnosticListener)
+        {
+            _logger = logger;
+            _diagnosticListener = diagnosticListener;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context context)
+        {
+            // These enabled checks are virtual dispatch and used twice and so cache to locals
+            var diagnosticsEnabled = _diagnosticListener.IsEnabled();
```

Comment:
This is a hot path (happens on every request whether logging is on or not), so we should try to minimize the amount of tests we make.   It is also nice if we use few local vars so that the prolog of the method is small.   Thus we would like at the top most level only three tests (one for diagnosticListener, one for ilogger logging and one for EventSource logging).    Everything else is under those three branches.   (thus when everything is off we do the minimum work).  Inside the diagnosticEnabled branch, we only test for Activity, and only test for the legacy if that FAILS.  That saves us a test in the ActivityCase (Which we expect to be common).